docs(container.orchestration): added docs for ContainerInstance enfor… (#5184)
eclipse/kura: docs/core-services/container-orchestration-provider-usage.md

- **Restart Container On Failure** - A boolean that tells the container engine to automatically restart the container when it has failed or shut down.

> - **Container Image Enforcement Digest** - A string representing the image digest allowed to be run from this container instances. It is used in the [Container Enforcement](#container-enforcement) service provided by the Container Orchestration Service.

Reviewer: ```suggestion
- **Container Image Enforcement Digest** - A string representing the digest for the image that will be allowed to run by this container instance. It is used in the [Container Enforcement](#container-enforcement) service provided by the Container Orchestration Service.
```

Maybe it is worth to provide an example of the string? Do we expect this to follow a certain pattern?

![Container Orchestration Provider Enfrocement Multiple Digests](./images/container-orchestration-provider-enforcement-multiple-digests.png)

## Container Instance Enforcement

> In the [Container Instance configuration](#configuring-the-container) is included an option that allows to add a specific digest to the enforcement allowlist: this option was designed to allow the authorization of a specific image, but only when a Container Instance run by Kura is using it.

Reviewer: ```suggestion
The [Container Instance configuration](#configuring-the-container) contains an option that allows to add a specific digest to the enforcement allowlist. When this is specified, only the matching image is allowed to execute. The framework will strictly enforce the policy to the containers running, in order to match a 1:1 relationship between running containers and defined container instances. 
```

> Everytime the enforcement feature performs a check on its startup or when a new container is started, it will consider as authorized all the digests included in the Container Orchestration Allowlist and those provided by the enabled Container Instances in Kura through the *Container Image Enforcement Digest* option.

Reviewer: Please review, it is not really understandable the statement

Author: Something like this would be ok?

> The [Container Instance configuration](#configuring-the-container) contains an option called *Container Image Enforcement Digest*, that allows to add a specific digest to the enforcement allowlist. When this is specified, only the matching image is allowed to execute. The framework will strictly enforce the policy to the containers running, in order to match a 1:1 relationship between running containers and defined container instances. 
> 
> Whenever Kura executes a *digest-scan* on the running containers, the digests provided in the Container Instances configurations are added to the Container Orchestration Allowlist: in this way, the authorized images will be not only the ones identified by the Allowlist digests, but also the ones specified through the *Container Image Enforcement Digest* options.

So the authorization schemas presented in the [previous section](#container-enforcement) are updated like:
![Enforcement Flow With Instances](./images/diagramAllowlistWithInstances.png)

> As you can see, the *ContainerInstance2* digest is not included in the final allowlist, because it is disabled: so, if a container with digest *DIGEST Y* is started under the image condition, it will then be stopped and deleted.

Reviewer: ```suggestion
As it can be seen, the *ContainerInstance2* digest is not included in the final allowlist, because it is disabled: so, if a container with digest *DIGEST Y* is started under the image condition, it will then be stopped and deleted.
```

What is the "image condition"?

> Finally, everytime a ContainerInstance is disabled, deleted or updated, the enforcement feature will perform a check on all the running containers. This is done because the enforcement allowlist may no longer include the previously provided digest or the latter may have been replaced with a new one: there may therefore be containers that were previously allowed, But now they are no longer and must be stopped and deleted.

Reviewer: Please review also this one

Reviewer: Still to review

![Container Orchestration Provider Enfrocement Multiple Digests](./images/container-orchestration-provider-enforcement-multiple-digests.png)

## Container Instance Enforcement

The [Container Instance configuration](#configuring-the-container) contains an option called *Container Image Enforcement Digest*, that allows to add a specific digest to the enforcement allowlist. When this is specified, only the matching image is allowed to execute. The framework will strictly enforce the policy to the containers running, in order to match a 1:1 relationship between running containers and defined container instances. 

> Whenever Kura executes a *digest-scan* on the running containers, the digests provided in the Container Instances configurations are added to the Container Orchestration Allowlist: in this way, the authorized images will be not only the ones identified by the Allowlist digests, but also the ones specified through the *Container Image Enforcement Digest* options.

Reviewer: ```suggestion
Whenever Eclipse Kura executes a *digest-scan* on the running containers, the digests provided in the Container Instances configurations are added to the Container Orchestration Allowlist: in this way, the authorised images will be not only the ones identified by the Allowlist digests, but also the ones specified through the *Container Image Enforcement Digest* options.
```

Whenever Eclipse Kura executes a *digest-scan* on the running containers, the digests provided in the Container Instances configurations are added to the Container Orchestration Allowlist: in this way, the authorised images will be not only the ones identified by the Allowlist digests, but also the ones specified through the *Container Image Enforcement Digest* options.

Everytime the enforcement feature performs a check on its startup or when a new container is started, it will consider as authorized all the digests included in the Container Orchestration Allowlist and those provided by the enabled Container Instances in Eclipse Kura through the *Container Image Enforcement Digest* option.

### Example Scenario

#### Container Instance Digest

Let's suppose to have a device on which Eclipse Kura is running with the Container Orchestration Service enabled, its Enforcement feature disabled and a Container Instance named `test-container` up and running on the system.

> An user wants to activate the Enforcement feature, so it enables the `Allowlist Enforcement Enabled` option in the Container Orchestration Service, but leaves the `Container Image Allowlist` option blank because it wants that no containers are started outside the framework (for details see the [Container Enforcement paragraph](#container-enforcement)).

Reviewer: ```suggestion
A user wants to activate the Enforcement feature, so it enables the `Allowlist Enforcement Enabled` option in the Container Orchestration Service, but leaves the `Container Image Allowlist` option blank because it wants that no containers are started outside the framework (for details see the [Container Enforcement paragraph](#container-enforcement)).
```

In order to allow the Container Instance, the user should provide the correct digest in the instance settings. Let's suppose that the digest of the image from which the container is created is `sha256:0000000000000000000000000000000000000000000000000000000000000000` and that the user fills the corresponding option with it: once the Container Instance is updated, the provided digest is included in the Enforcement feature allowlist, so the container will be allowed to run without interference. This case can be summarised as:

![Container Instance Schema Success](./images/container-instance-enforcement-success.png)

Once the instance digest is added to the enforcement feature, it can be used also to authorise container run by the Command Line Interface, or other instances on Eclipse Kura without providing the digest option. But what happen if the Container Instance is disabled? The aim of the Container Instance Enforcement Digest is to be used as an authorization method of the instance itself: this means that its digest is added to the enforcement feature allowlist **only if the instance is enabled**.

![Schema with disabled instances](./images/schemaWithoutInstances.png)

> As it can be seen from the image, the merged `Enforcement Allowlist` box doesn't contain the digest associated to the Container Instance, because, being it disabled, the corresponding digest is ignored. If the enforcement is enable and a container with digest *DIGEST Z* is started, it will then be stopped and deleted, because its digest won't be included in the allowlist.

Reviewer: ```suggestion
As it can be seen from the image, the merged `Enforcement Allowlist` box doesn't contain the digest associated to the Container Instance, because, being it disabled, the corresponding digest is ignored. If the enforcement is enabled and a container with digest *DIGEST Z* is started, it will then be stopped and deleted, because its digest won't be included in the allowlist.
```

> Finally, everytime a *Container Image Enforcement Digest* option is modified, or the ContainerInstance is disabled or deleted, the enforcement feature will perform a check on all the already running containers. This is done because if the digest that was previously provided has changed after an instance update, or removed due to disabling or deleting the instance, those containers that were previously authorised by this digest are no longer allowed to run. So they must be stopped and deleted.

Reviewer: ```suggestion
Finally, everytime a *Container Image Enforcement Digest* option is modified, or the ContainerInstance is disabled or deleted, the enforcement feature will perform a check on all the already running containers. This is done because, if the digest that was previously provided has changed after an instance update, or removed due to disabling or deleting the instance, those containers that were previously authorised by this digest are no longer allowed to run. So they must be stopped and deleted.
```

Whenever Eclipse Kura executes a *digest-scan* on the running containers, the digests provided in the Container Instances configurations are added to the Container Orchestration Allowlist: in this way, the authorised images will be not only the ones identified by the Allowlist digests, but also the ones specified through the *Container Image Enforcement Digest* options.

Everytime the enforcement feature performs a check on its startup or when a new container is started, it will consider as authorized all the digests included in the Container Orchestration Allowlist and those provided by the enabled Container Instances in Eclipse Kura through the *Container Image Enforcement Digest* option.

So the authorization schemas presented in the [previous section](#container-enforcement) are updated like:
![Enforcement Flow With Instances](./images/diagramAllowlistWithInstances.png)

> As you can see from the image, the final `Enforcement Allowlist` box doesn't contain the *ContainerInstance2*, because, being it disabled, the corresponding digest is ignored: so, if the enforcement is enable and a container with digest *DIGEST Y* is started, it will then be stopped and deleted, because its digest won't be included in the allowlist.

Reviewer: ```suggestion
As it can be seen from the image, the final `Enforcement Allowlist` box doesn't contain the *ContainerInstance2* <what?>, because, being disabled, the corresponding digest is ignored. If the enforcement is enabled and a container with digest *DIGEST Y* is started, it will then be stopped and deleted, because its digest won't be included in the allowlist.
```

I still think that this sentence is a bit convoluted and not fully understandable
